### PR TITLE
Jest

### PR DESCRIPTION
**Pull request by astridx**
See https://github.com/maplibre/maplibre-gl-js/issues/273

### CODE REVIEW

**Comment by HarelM** [comment]
This is probably a good start.
Do we really need babel? I've read somewhere there's something called `ts-jest`, can we use that?
Did you convert the test manually or did you use a tool of some sort?
The reference of the tests is pointing to the build output (..... rollup/...) and needs to point to the typescript files, I guess...?
I would prefer to see this change as one file removed and one file added kind of thing and maybe git will consider it as a rename commit with some changes as this would not allow to track what was converted easily, I think...

In any case thanks for pushing this forward! :-)

**Comment by astridx** [comment]
> Do we really need babel? I've read somewhere there's something called `ts-jest`, can we use that?

If i see it correctly we need babel if we use the version in the rollup directory, as i suggest here, because we test a js file in the rollup directory. 

> Did you convert the test manually or did you use a tool of some sort?

I converted the three example tests manually.

> The reference of the tests is pointing to the build output (..... rollup/...) and needs to point to the typescript files, I guess...?

The current tests also use the rollup directory: https://github.com/maplibre/maplibre-gl-js/blob/39579f7e3ba8e5a9e9e37b379dff92744047a4c3/test/unit/mapbox-gl.test.js#L3
I like this. This helps us ensure that build tools don’t introduce any regressions. Besides, otherwise we would have to transform glsl for the tests and there could also be an inconsistency.

> I would prefer to see this change as one file removed and one file added kind of thing and maybe git will consider it as a rename commit with some changes as this would not allow to track what was converted easily, I think...

For starters, I would prefer `jest `to be created alongside to `tape`. Because that way we can build up experience first.

**Comment by HarelM** [comment]
The tests are currently written in javascript so it made sense to point them to javascript output, I don't think this is the right approach when dealing with typescript.
The part about seems to me to suffer from the same problem.
My 2 cents are: the more we use standard tools and their standard usage the better we are off.
So unit tests should be in typescript and reference typescript code.
We have other tests to make sure the output is correct, runs in the browser well etc...

The current state is far from ideal - where we test the output code. This is why I opened this issue in the first place.
The current situation is simply a temporary state to allow to continue development without holding back PR due to the transition...

**Comment by wipfli** [comment]
Thanks for working on this Astrid! Looking forward to more modern testing.

**Comment by astridx** [comment]
> The tests are currently written in javascript so it made sense to point them to javascript output, I don't think this is the right approach when dealing with typescript.

You are certainly right. Many tests will also become unnecessary this way, because with typscript incorrect entries no longer need to be tested in many cases. Because these are not possible.


> The part about seems to me to suffer from the same problem.

I don't understand that sentence. If it is important, please explain it to me.


> My 2 cents are: the more we use standard tools and their standard usage the better we are off.
> So unit tests should be in typescript and reference typescript code.
> We have other tests to make sure the output is correct, runs in the browser well etc...

I have now changed  everything so that the typescript-files in the `src `directory are imported into the few sample tests. I have also changed the` root directory` for the `jest tests` to `src`. This way the tests can be stored next to the actual files, which is what you suggested in the issue https://github.com/maplibre/maplibre-gl-js/issues/273#issuecomment-914577636.



> Thanks for working on this Astrid! Looking forward to more modern testing.

:)


Do you like this set-up of the environment more? The actual work can only begin once we have defined it.

**Comment by HarelM** [comment]
This looks great on terms of setup!!
Lots of typos in the readme :-) but great explanation.
I have a problem that I can't compare the tests that were made before the migration to the tests made after the migrating because all tests are new.
How world you recommend I review the tests?
Also worth trying is this repo to automate the migration somehow to reduce missing tests:
https://github.com/skovhus/jest-codemods
I don't know.
In general, looks like a promising start...

File: docs/tests.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+Snapshots:   0 total
+Time:        15.98 s, estimated 20 s
+Ran all test suites.
+```
+
+#### Configuration
+
+The configuration of Jest is defined by the file `jest.config.js`. 
+
+##### preset
```
Comment (HarelM) [inline docs/tests.md:34]: I think the part below is redundant - saying that the configuring of the tests are in jest.config.js file is enough. The dude is documenting itself in the config and toy can place the relevant comments there if needed...
This is what I feel at least...

Comment (astridx) [reply docs/tests.md]: You are right. There is a link to the jest documentation at the beginning of the config file. That is enough. This also prevents outdated information from appearing there. I deleted this part.

**Comment by astridx** [comment]
> I have a problem that I can't compare the tests that were made before the migration to the tests made after the migrating because all tests are new.
> How world you recommend I review the tests?


Do you think these steps are a possible way to go on:

1. The directory `test/unit_copy` contains a copy of each test in the directory `test/unit`. The files are already transformed using https://github.com/skovhus/jest-codemods. Select a test file and move it into the `src` directory. For example, move the file `test/unit_copy/geo/lng_lat.test.js` to `src/geo/lng_lat.test.js`.

2. Change the file extension from `js` to `ts`.

3. Correct the imports so that the source type script file is used. For example, change `import LngLat from '../../../rollup/build/tsc/geo/lng_lat'` to `import LngLat from '.lng_lat'`.

4. Check the transformation from Tape to Jest. Information can be found at https://jestjs.io/docs/getting-started.

5. Correct errors due to the change from JavaScript to Typesript.

6. Check via `npm run test-unit` if the test is executed without errors.

7. Run `npm run lint` to make sure the coding style fits.

8. Create a Pull Request containing your changes. If you have migrated one file, then the Pull Request includes at least the deletion of the file in directory `test/unit_copy/**` and the addition of the file in directory `src/**`.


I have transformed one file as an example. You should compare the files `test/unit/geo/lng_lat.test.js` to `src/geo/lng_lat.test.js`

> Also worth trying is this repo to automate the migration somehow to reduce missing tests:
> https://github.com/skovhus/jest-codemods

I have copied all unit tests and manipulated the copies with jest-codemods. I'm not sure if it's more confusing. But the real unit tests are still there.

**Comment by HarelM** [comment]
This is great in terms of work methodology, and will make sure we convert all the tests, but it doesn't fill the need of someone that would like to review this PR.
When I come to this PR and I would like to review it, I don't have a reference to the original test file in order to compare it and understand the relevant changes.
Of course I can open the code and look for it but that beats the purpose.
My question is if there's a way to make git understand the relevant changes like we had in the typescript branch, I.e:
1. Convert the tests from tap to jest in a single commit (CI will be fail at that point)
2. Convert the tests from js to ts (rename extension is enough? Can we use a tool here?)
3. Start fixing the tests files until CI passes
4. Review the changes file by file

It's different than what @wipfli said because it's not gradual, but if we what to make it gradual we probably need a branch that we convert one file or more at a time.
I don't know...
Bottom line, I think copying the tests didn't allow to review changes well...

**Comment by wipfli** [comment]
For easier reviewing I think we need a clean history and we review commit after commit, even if some of them fail CI. First commit could be automatic conversion. Second could be move the unit test from folder a to b and fix imports and conversion problems.

**Comment by wipfli** [comment]
Thanks Astrid for pushing this forward.

**Comment by HarelM** [comment]
Commit by commit can be a good solution as well, but I think keeping the old tests is not needed - i.e. when migrating a test delete the old one...
There's also a need to see what part of the `stub_loader` is relevant as it seems that jest has some support of jsdom (I haven't read about it but just the config part saying use jsdom seems like it might be that some of the code there is not needed, maybe, maybe not).

**Comment by astridx** [comment]
I am closing this branch because it contains too much confusion. If it is possible to submit this in a more tidy way, I will do so. If someone else is quick, that's great.
